### ISSUE TITLE
RxSwift interop

### ISSUE THREAD

**Issue by tassilo-posegga** [badoo/Reaktive#538]
Any chance to see RxSwift interop, similar to what we have with RxJava?

**Comment by arkivanov** [comment]
Thanks for raising the issue! Indeed it would be cool to have such interop. But it is tricky because of the Objective-C layer and its [limitations](https://kotlinlang.org/docs/reference/native/objc_interop.html). I will keep this issue open and we'll investigate possibilities.

**Comment by arkivanov** [comment]
@mikezliu asked in #551 for:

> Similar to the java equivalents (i.e. asRxJava3Single, asRxJava3Observable), it would be nice to have these extensions for swift

**Comment by arkivanov** [comment]
While we are working on this, here is a copy-paste workaround (Reaktive -> RxSwift interop):

```swift
extension Observable where Element : AnyObject {
    static func from(_ reaktiveObservable: ReaktiveObservableWrapper<Element>) -> Observable<Element> {
        return Observable<Element>.create { observer in
            let disposable = reaktiveObservable.subscribe(
                isThreadLocal: false,
                onError: { observer.onError(KotlinError($0)) },
                onComplete: observer.onCompleted,
                onNext: observer.onNext
            )
            
            return Disposables.create(with: disposable.dispose)
        }
    }
}

extension Single where Element : AnyObject {
    static func from(_ reaktiveSingle: ReaktiveSingleWrapper<Element>) -> Single<Element> {
        return Single<Element>.create { observer in
            let disposable = reaktiveSingle.subscribe(
                isThreadLocal: false,
                onError: { observer(.failure(KotlinError($0))) },
                onSuccess: { observer(.success($0)) }
            )
            
            return Disposables.create(with: disposable.dispose)
        }
    }
}

extension Maybe where Element : AnyObject {
    static func from(_ reaktiveMaybe: ReaktiveMaybeWrapper<Element>) -> Maybe<Element> {
        return Maybe<Element>.create { observer in
            let disposable = reaktiveMaybe.subscribe(
                isThreadLocal: false,
                onError: { observer(.error(KotlinError($0))) },
                onComplete: { observer(.completed) },
                onSuccess: { observer(.success($0)) }
            )
            
            return Disposables.create(with: disposable.dispose)
        }
    }
}

extension Completable {
    static func from(_ reaktiveCompletable: ReaktiveCompletableWrapper) -> Completable {
        return Completable.create { observer in
            let disposable = reaktiveCompletable.subscribe(
                isThreadLocal: false,
                onError: { observer(.error(KotlinError($0))) },
                onComplete: { observer(.completed) }
            )
            
            return Disposables.create(with: disposable.dispose)
        }
    }
}

struct KotlinError : Error {
    let throwable: KotlinThrowable
    
    init (_ throwable: KotlinThrowable) {
        self.throwable = throwable
    }
}
```

**Comment by arkivanov** [comment]
Spent today trying to add an RxSwift interop module. At the moment it appears there is no convenient way to add and publish such a module. The reason is as follows.

We can create an Xcode Framework project wich will depend on both RxSwift and Reaktive frameworks. As a result there will be a new framework (e.g. `rxswift-interop`). 
Next on client side we have some `shared` module that depends on Reaktive. This `shared` module produces `shared` framework with Reaktive exported. Now we have an Xcode app `client-app`, which uses RxSwift. And we are adding both `shared` and `rxswift-interop` frameworks to `client-app` as well. 
If try to use an extension from `rxswift-interop` framework on an `Observable` from `shared` framework, then there is type mismatch error.

We will think about this case, but at the moment it is not promising. In the worst case we will just publish some "copy-paste" code snippets.

**Comment by cl3m** [comment]
@arkivanov How do you get ReaktiveObservableWrapper<T> in the swift header? I have only ReaktiveObservable without generic type. 

Btw did you make any progress on rxswift-interop ?

**Comment by arkivanov** [comment]
@cl3m My last comment is still valid, due to the fact how Kotlin/Native is exported to Swift currently it looks impossible to create and publish any consumable interop. We will likely provide some copy-paste solution soon.

**Comment by arkivanov** [comment]
So the original issue why this was considered as not possible is the type incompatibility. If there is let say a `core` Kotlin module, and two Kotlin modules exported as separate frameworks `feature1` and `feature2`, then we can't pass `core` objects between `feature1` and `feature2. Both frameworks will have their own copies of the `core` module and own runtimes.

But it seems there is a workaround that we can use to actually create and publish a Swift interop library. Instead of the `ObservableWrapper` Kotlin class we can expose simple function types. They will be converted to standard Obj-C types, and there will be no issues with type compatibility.

I bumped into an example - [KMP-NativeCoroutines](https://github.com/rickclephas/KMP-NativeCoroutines) is a library providing interop for Kotlin coroutines and RxSwift/Combine. It exports coroutine types as [function typealiases](https://github.com/rickclephas/KMP-NativeCoroutines/blob/205488b6314b27fde1779d856784bc616e4f8018/kmp-nativecoroutines-core/src/appleMain/kotlin/com/rickclephas/kmp/nativecoroutines/NativeFlow.kt#L17). And here an example of the the [Swift part](https://github.com/rickclephas/KMP-NativeCoroutines/blob/master/KMPNativeCoroutinesRxSwift/Observable.swift).

We should investigate it further.

**Comment by arkivanov** [comment]
So I investigated this further and there is a workaround indeed. It is possible to create a Swift interop framework, however the API of exported things will not be as good as with wrapping (`ReaktiveObservableWrapper`, etc.).

The biggest limitation is that such a Swift interop framework will not be able to depend on Reaktive and use its types. This is because Reaktive types are not compatible when such a Swift framework is used in a Swift project, which also depends on Reaktive. This in turn is because the Swift interop framework and the Swift project will have different copies of Reaktive.

The workaround is to not depend on Reaktive and use only Cocoa types. 

In Kotlin we can define the following (just an example for `Observable`):

```kotlin
typealias ExportedCancellable = () -> Unit

expect class ExportedError

actual typealias ExportedError = NSError // For Darwin targets
actual typealias ExportedError = Throwable // For all other targets

typealias ExportedObservable<T> =
        (
        onNext: (T) -> Unit,
        onComplete: () -> Unit,
        onError: (ExportedError) -> Unit
    ) -> ExportedCancellable

fun <T : Any> Observable<T>.export(): ExportedObservable<T> =
    { onNext, onComplete, onError ->
        subscribe(
            onNext = onNext,
            onComplete = onComplete,
            onError = { onError(it.export()) }
        ).export()
    }

private fun Disposable.export(): ExportedCancellable = ::dispose

fun <T : Any> ExportedObservable<T>.import(): Observable<T> =
    observable { emitter ->
        emitter.setCancellable(
            invoke(
                emitter::onNext,
                emitter::onComplete,
                { emitter.onError(it.import()) }
            )
        )
    }
```

This can be used in Kotlin as follows:
```kotlin
class SharedLogic {
    val strings: ExportedObservable<String> =
        observableOf("Str1", "Str2")
            .subscribeOn(computationScheduler)
            .observeOn(mainScheduler)
            .export()

    fun accept(strings: ExportedObservable<String>) {
        strings
            .import()
            .subscribe {
                // Handle values
            }
    }
}
```

Here is the Swift side
```swift
func foo() {
    let logic = SharedLogic()
    
    RxSwift.Observable
        .from(logic.strings, unit: KotlinUnit()) // Using the extension from the interop framework
        .subscribe {
            // Handle values
        }
}

func bar() {
    let logic = SharedLogic()
    
    logic.accept(
        strings: RxSwift.Observable
            .from(["Str2", "Str2"])
            .asReaktive(unit: KotlinUnit()) // Using the extension from the interop framework
    )
}
```

Notice that you have to pass `KotlinUnit` to the Swift interop extensions. This is due to the following limitation - the Kotlin function `(() -> Unit) -> Unit` is mapped to Swift function `(() -> KotlinUnit) -> Void`. Because the Swift interop can't use Kotlin library, `KotlinUnit` has to be provided from outside.

On the Android side we also have to export/import things:
```kotlin
fun foo() {
    val logic = SharedLogic()

    logic.strings.import().subscribe {
        // Handle values
    }
}

fun bar() {
    val logic = SharedLogic()

    logic.accept(
        observableOf("Str1", "Str2")
            .export()
    )
}
```

Or just expose separate non-exported methods/properties in `SharedLogic`.

**I have a strong feeling the the current approach with wrappers and the requirement to copy-paste a bit works just fine, and we can avoid messing with all this interop issues. Could somebody provide a feedback, what do you thing?**

**Comment by rickclephas** [comment]
@arkivanov I have been thinking about this as well. For the coroutines library it isn't much of an issue to pass the `Unit` type since it is all hidden inside the library. Though there are other cases where that might not be the case.
There seems to be another way to accomplish this. Instead of using Kotlin types use ObjC types.
I haven't really tested this but have been [playing](https://github.com/rickclephas/KMP-Logging) with it a little.
So far it looks promising (e.g. it compiles 🙃). Though it does make it harder to write a library this way.

**Comment by arkivanov** [comment]
@rickclephas I tried to hide Unit passing inside the library and it worked. However we need two-way interop: `Swift -> Reaktive-> Swift`. And so either we still need to supply `KotlinUnit` from outside for one of the cases, or define two different typealiases. I found the API cleaner, when there is single set of typaliases and passing `KotlinUnit` from outside.

**Comment by rickclephas** [comment]
@arkivanov alright I am not sure about the specifics for Reaktive but I think it might work.
The limitation we currently have is that Kotlin Native types are different between modules meaning class A in library B isn't the same type as the same class A in library C. 
So I was thinking how we could overcome that limitation and came up with the following (theory).
Instead of defining the API in Kotlin we define it in Swift/ObjC and use the expect/actual feature to use the Swift types for Kotlin Native instead of pure Kotlin types. This does however require some work to make sure the defined Swift API matches with the Kotlin API 😅.

**Comment by arkivanov** [comment]
Thanks @rickclephas for your input, I will think if we can apply it.

**Comment by arkivanov** [comment]
Alright, the idea to only use ObjC types pushed me to the following workaround.

In Kotlin:
```kotlin
typealias ExportedCancellable = () -> Any?

typealias ExportedObservable<T> =
        (
        onNext: (T) -> Any?,
        onComplete: () -> Any?,
        onError: (ExportedError) -> Any?
    ) -> ExportedCancellable
```

In Swift interop:
```swift
public typealias ReaktiveCancelable = () -> Any?

public typealias ReaktiveObservable<T> =
    (
        _ onNext: @escaping (T) -> Any?,
        _ onComplete: @escaping () -> Any?,
        _ onError: @escaping (Error) -> Any?
    ) -> ReaktiveCancelable
```

Usage of the interop framework in Swift:
```kotlin
func foo() {
    let logic = MyLogic()

    RxSwift.Observable
        .from(logic.strings) // RxSwift extension from the interop framework
        .subscribe { NSLog($0) }

    logic.strings(
        obs: RxSwift.Observable
            .from(["Str2", "Str2"]) // RxSwift extension from the interop framework
            .asReaktive()
    )
}
```

In this case we can just return `null`/`nil` from functions, no need to deal with `KotlinUnit`.

But I'm still not convinced to add such an API, I still believe that what we currently have is better and cleaner, despite the need to copy-paste the [interop code](https://github.com/badoo/Reaktive/blob/master/docs/SwiftInterop.md#rxswift-interop) (only once per project).

So you just copy the following interop code for e.g. `Observable`:
```swift
extension RxSwift.Observable where Element : AnyObject {
    static func from(_ observable: ObservableWrapper<Element>) -> RxSwift.Observable<Element> {
        return RxSwift.Observable<Element>.create { observer in
            let disposable = observable.subscribe(
                isThreadLocal: false,
                onError: { observer.onError(KotlinError($0)) },
                onComplete: observer.onCompleted,
                onNext: observer.onNext
            )

            return Disposables.create(with: disposable.dispose)
        }
    }
}
```

And then in Kotlin you just `wrap`:
```kotlin
class SharedDataSource {
    fun load(): ObservableWrapper<String> =
        observable { emitter ->
            // A long running operation
        }
            .subscribeOn(ioScheduler)
            .observeOn(mainScheduler)
            .wrap()
}
```

Wrappers are normal Reaktive types (e.g. `ObservableWrapper` class implements `Observable` interface), and so they can be used **directly** on Swift/Android/JavaScript/etc. Or converted to RxSwift/RxJava/etc. as shown above when needed.

**Comment by arkivanov** [comment]
Based on the information above, both ways (the current with Wrappers and the discovered new one) have pros and cons. Seems like Wrappers is a decent solution, let's continue use it for now. Closing the issue until there are additional circumstances or requests from the community.

**Comment by arkivanov** [comment]
There is a new Gradle plugin [MOKO KSwift](https://github.com/icerockdev/moko-kswift) which looks very promising. Seems like it provides an API for custom Swift code generation.
